### PR TITLE
fix(core,runtime): make WDDM residency budgeting opt-in on desktop-shared GPUs

### PR DESCRIPTION
**Pull request by keylimesoda**
## Summary

On an RTX 4090 that is **also driving the Windows desktop**, v1.0.0 and v1.1.0 load the model successfully and then emit exactly one garbage character per token. v0.9.0 is unaffected with the same artifact.

Sample output from v1.1.0 (`ninfer-rtx4090-windows-x64-v1.1.0.zip`, official `qwen3_8_27b.ninfer`, sha256 `eec39564...`):

```json
{"content":"","reasoning_content":"/2*/(&!0!-*()'#+#)%%!!*+0!\"'02\",2$2(\"'+2"}
```

The tell is that `reasoning_content` is **exactly one character per token** (`predicted_n: 40` → 40 characters), and every character falls in the contiguous ASCII range `0x21`–`0x33`. A model producing incoherent text would still emit real vocabulary, so this is a memory problem rather than a numerical one.

Bisecting `v0.9.0-rtx4090..v1.0.0-rtx4090` points at a35acf6a (WDDM residency).

## Root cause

a35acf6a budgets runtime capacity as

```cpp
return std::max(free_after_weights, total_bytes - weight_bytes - kMinDwmHeadroom);
```

and then eagerly commits the entire arena, on the assumption that WDDM will evict every other allocation on the adapter down to a 512 MiB DWM floor.

When the GPU is also scanning out a desktop, those background allocations are frequently **not** evictable. On my machine (~4 GiB of desktop VRAM in use across the 4090, an AMD iGPU and a Remote Display Adapter) the engine reported `free-after-weights=7.57 GiB` — exactly `23.99 − 15.92 − 0.5` — while roughly 2.7 GiB was genuinely available. The runtime oversubscribes VRAM and the served output is corrupt.

This is invisible on a card that isn't driving a display, which I assume is why it shipped.

## Changes

- **Gate the residency lock, the eager full-arena commit, and the total-VRAM budget behind `NINFER_WDDM_EVICTABLE_BUDGET=1`.** Default returns to the pre-a35acf6a behaviour: size the arena from measured free memory and let `cudaMalloc` commit lazily. Headless cards can opt back in and keep the larger context ceilings.
- **Resolve KV capacity against the budget the arena was actually sized from.** The arena owns that memory, so re-measuring with `cudaMemGetInfo` afterwards reports ~0 free and the capacity gate rejects otherwise-valid configurations. (This is what the `max()` was implicitly compensating for.)

Two genuine defects in the D3D12 path are fixed as well. **Neither is sufficient to explain the corruption on its own** — with the residency path force-enabled via the new env var, output is still wrong — but both are API misuse:

- `EnqueueMakeResident` is asynchronous and signals the supplied fence once the heap is resident. The heap was mapped into CUDA and written before that fence was ever waited on. This now waits, and falls back to `cudaMalloc` on timeout rather than running against non-resident memory.
- `cudaExternalMemoryDedicated` applies to dedicated allocations such as D3D12 committed resources, not to suballocatable D3D12 heaps, so it is dropped when importing the heap.

## Verification

RTX 4090 shared with the Windows desktop, CUDA 13.3, VS 2022, `CMAKE_CUDA_ARCHITECTURES=89`.

| Build | Result |
|---|---|
| v1.0.0 released binary | garbage |
| v1.0.0 built from source, unpatched | garbage (control) |
| Same tree + this change | correct text, `finish_reason: "stop"` |
| v1.1.0 + this change, `rk4v4-e8` + MTP4 @ 98304 ctx | correct, **78.7 tok/s** |

The A/B was single-variable: identical tree and toolchain, only this patch differing.

After the change, accounting is honest again — `free-after-weights=5.83 GiB`, `free-after-startup=2.78 GiB` — and the DirectStorage prompt cache works:

```
[req 2] done finish=stop_token prompt=68 gen=370 cache=66
        reuse=restore_turn_checkpoint ttft=249ms decode=96.9tok/s
```

`ctest`: **83/84 pass.** `ninfer_qwen3_6_frontend_test` fails identically with and without this change (`unsupported frontend/chat_template.jinja`), so it is pre-existing and unrelated — I verified by stashing the patch and re-running.

## Note on the README context table

The published ceilings (352k for `rk4v4-e8`+MTP4, 462k for `rk2v4-e8`) appear to have been measured with the optimistic budget on a card not driving a desktop. With honest accounting and ~4 GiB of desktop VRAM in use, this machine reaches ~98k. That is a property of the measurement environment rather than anything this patch changes, but it may be worth a caveat in the table.

Happy to adjust the approach — for example defaulting the env var on and detecting whether the adapter has an attached display instead — if you'd prefer to keep the residency path enabled by default.

### CODE REVIEW

**Comment by keylimesoda** [comment]
Follow-up: I dug further into *why* the residency path fails, since my original writeup could only say "not sufficient on its own". Two new data points, both from this machine (RTX 4090 shared with the Windows desktop).

**1. The D3D12 mapping itself is fine.**

I added a read-back check at allocation time (second commit): write a pattern to the head and tail of the arena through the CUDA mapping, sync, read it back, compare. With `NINFER_WDDM_EVICTABLE_BUDGET=1` the check **passes** — the imported heap round-trips correctly. So the corruption is not a bad import, a wrong offset, or a LUID mismatch. The memory is valid when handed out.

**2. The failure is a stall, not just corruption.**

With the residency path enabled at only **16384** context — weights 15.92 GiB + runtime 1.05 GiB against 24 GiB, roughly 4 GiB spare, so *not* oversubscribed — the server loads weights and then never reaches KV capacity resolution. It sits burning **exactly 1.0 s of CPU per wall-clock second on one core with the GPU at 7-9% utilization**:

```
t=15s cpu=96.5s  gpu=20284 MiB, 7 %
t=30s cpu=111.5s gpu=20284 MiB, 9 %
t=45s cpu=126.5s gpu=20284 MiB, 7 %
t=60s cpu=141.5s gpu=20284 MiB, 8 %
```

That is CUDA spin-waiting on work that is not progressing, which is what WDDM paging looks like from the CPU side.

So the picture is: the heap imports cleanly and verifies, but operating on it is pathological even with headroom to spare. My guess is that `D3D12_RESIDENCY_FLAG_DENY_OVERBUDGET` plus `RESIDENCY_PRIORITY_MAXIMUM` interacts badly when another session's compositor holds allocations on the same adapter — this box additionally has an active console session alongside the RDP session I was testing from, so there are two DWMs in play. I have not proven that mechanism, and I want to be clear it is a hypothesis rather than a finding.

What this does confirm is that the defaults in the PR are the right shape: the residency scheme needs to be opt-in until it can verify it actually got what it asked for, because every failure mode I have hit is silent.

Happy to test any instrumentation you'd like on this configuration — it seems to reproduce reliably here, which may be more useful than the patch itself.

**Comment by UDPSendToFailed** [comment]
Hi!

Thank you for the PR and testing it out. Since I only test the builds on my personal setup (RTX 4090 as the primary GPU, 4K 240Hz monitor connected via DP, 64GB system RAM), I haven't noticed this issue. The goal with this WDDM "hack" was to free up VRAM by having Windows page out other applications into the system memory, which now I see isn't really working on every setup universally as it did on mine. 

I think the approach of having it disabled by default and leaving it as opt-in is a good idea, so it's more like an experimental feature than an enforced fragile default. One thing I'd add to it, I would prefer keeping the toggle as a CLI argument rather than an environment variable for consistency with the rest of the options in the codebase.

**Comment by Suffice** [comment]
@keylimesoda great catch with the round-trip probe — that read-back verification is exactly what this code needed. One heads-up from my side: after the probe, there are still two fills with no ordering against non-blocking streams, and I hit a nasty silent bug through one of them on an RTX 4090 (constant token-0 output, every sanitizer clean).
**What's left:**
1. The **final full-capacity zero-fill** at the end of `try_allocate_d3d12_residency_locked` — the probe's `cudaDeviceSynchronize()` drains the *first* touch, but this last fill is enqueued on the legacy default stream and the function returns without draining it.
2. The **fallback `cudaMalloc` path's touch fill** in `DeviceArena::DeviceArena` has no sync at all.
Since the legacy default stream does not order against non-blocking streams like `DeviceContext::load_stream`, async H2D weight copies can land *before* either of these fills and get wiped to `0x00`. Failure mode: all-zero device weights → uniform logits → the model emits token 0 for every prompt, load stats look perfect (16.67 GiB / 100%), and compute-sanitizer memcheck/racecheck/synccheck are all clean — it's an ordering race between two legal operations, not a memory error.
**Minimal fix:** `cudaDeviceSynchronize()` after the final D3D12 zero-fill (failure path frees the imported external memory) and after the fallback touch — 19 insertions / 2 deletions in `src/core/arena.cu`, plus a `PinnedHostBuffer` write-combined → `cudaMallocHost` swap I verified alongside. It's on a branch right on top of your merge: https://github.com/Suffice/ninfer-4090/tree/fix/wddm-touchfill-sync-followup — happy to open a PR from it if you'd like, or cherry-pick/adapt as you see fit.
**Verified:** Windows 11 · RTX 4090 · CUDA 13.3 — `ninfer_artifact_materialization_test` passes (it read back `0x00` where artifact data belongs before the fix), qwen3_8_27b generates coherent output at ~35 tok/s decode.
